lib/recursive-methods.rb

```diff
@@ -1,49 +1,92 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError.new "N must be greater than or equal to 0" if n < 0
+  return 1 if n <= 1
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return "" if s.empty?
+  return s[0] if s.length == 1
+  
+  return reverse(s[1..-1]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return helper_reverse(s)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def helper_reverse(s, start_index = 0, finish_index = (s.length - 1))
+  return s if start_index >= finish_index
+  temp = s[start_index]
+  s[start_index] = s[finish_index]
+  s[finish_index] = temp
+  return helper_reverse(s, start_index + 1, finish_index - 1)
+end
+
+# Time complexity: O(log(n))
+# Space complexity: O(log(n))
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0
+  return n + bunny(n / 2.0 )
+  # This passes the tests, though technically it never hits 0 (just approaches 0)
+  # An alternate solution is to return 1 + bunny(n - 0.5) but that would have a slower time complexity O(n)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def nested(s)
```

Comment:
👍  This works, but you have similar time/space issues with the `reverse` method.